src/runtime/classobj.cpp

```diff
@@ -735,6 +735,69 @@ static Box* instanceNext(BoxedInstance* inst) {
     return r;
 }
 
+static PyObject* slot_nb_power(PyObject* self, PyObject* other, PyObject* modulus) noexcept {
+    PyObject* func, *res;
+    PyObject* args;
+
+    static BoxedString* pow_str = internStringImmortal("__pow__");
+    if (modulus == Py_None) {
+        // Binary power
+        int do_other = Py_TYPE(self) != Py_TYPE(other) && other != None && Py_TYPE(other)->tp_as_number != NULL;
+        // Tenary power doesn't use __rpow__.
+        static BoxedString* rpow_str = internStringImmortal("__rpow__");
+
+        if (!do_other && (func = instance_getattro(self, pow_str)) != NULL) {
+            args = PyTuple_Pack(1, other);
+        } else {
+            PyErr_Clear();
+            if (other->cls == instance_cls) {
+                if ((func = instance_getattro(other, rpow_str)) != NULL) {
+                    args = PyTuple_Pack(1, self);
+                }
+            } else {
+                Box* func = _instanceGetattribute(self, rpow_str, false);
+                if (!func)
+                    return NotImplemented;
+                return runtimeCall(func, ArgPassSpec(1), other, NULL, NULL, NULL, NULL);
+            }
+        }
+    } else if ((func = instance_getattro(self, pow_str)) != NULL) {
+        // Tenary power
+        args = PyTuple_Pack(2, other, modulus);
+    }
+
+    if (func == NULL) {
+        if (!PyErr_ExceptionMatches(PyExc_AttributeError))
+            return NULL;
+        PyErr_Clear();
+        PyErr_Format(TypeError, "unsupported operand type(s) for ** or pow(): '%s' and '%s'", getTypeName(self),
+                     getTypeName(other));
+        return NULL;
+    }
+
+    res = PyEval_CallObject(func, args);
+    Py_DECREF(func);
+    return res;
+}
+
+static PyObject* slot_nb_inplace_power(PyObject* self, PyObject* arg1, PyObject* arg2) {
+    PyObject* func, *res;
+    PyObject* args;
+    args = PyTuple_Pack(1, arg1);
+    static BoxedString* ipow_str = internStringImmortal("__ipow__");
+    if ((func = instance_getattro(self, ipow_str)) == NULL) {
+        if (!PyErr_ExceptionMatches(PyExc_AttributeError))
+            return NULL;
+        PyErr_Clear();
+        PyErr_Format(TypeError, "unsupported operand type(s) for ** or pow(): '%s' and '%s'", getTypeName(self),
+                     getTypeName(arg1));
+        return NULL;
+    }
+    res = PyEval_CallObject(func, args);
+    Py_DECREF(func);
+    return res;
+}
```

Comment:
I'm not sure I understand the role of these functions -- they seem named after the typeobject.c functions, but 
behave a bit more like instance_pow.  Regardless, maybe in this situation we should just copy instance_pow and instance_ipow from CPython?  Your slot_nb_power seems reasonable (though a bit hard to verify all the corner cases); I think inplace_power is missing the fallback to `__pow__` and `__rpow__`.